Chap_Terms.tex

```diff
@@ -122,6 +122,12 @@ \section{Procedure Conventions}
 The call may both use and update the argument.
 \end{itemize}
 
+Many \ac{PMIx} interfaces, particularly nonblocking interfaces, use a \code{void*}cbdata object passed to the function that is then passed to the associated callback. In a client-side \ac{API}, the cbdata is a client-provided context (opaque object) that the client can pass to the nonblocking call (e.g., \refapi{PMIx_Get_nb}). When the nonblocking call (e.g., \refapi{pmix_value_cbfunc_t}) completes, the cbdata is passed back to the client without modification by the \ac{PMIx} library, thus allowing the client to associate a context with that callback. This is useful if there are many outstanding nonblocking calls.
+
+A similar model is used for the server module functions (see \ref{server:module_fns}). In this case, the \ac{PMIx} library is making an upcall into its host via the \ac{PMIx} server module function and passing a specific cbfunc and cbdata. The \ac{PMIx} library expects the host to call the cbfunc with the necessary arguments and pass back the original cbdata upon completing the operation. This gives the server-side \ac{PMIx} library the ability to associate a context with the call back (since multiple operations may be outstanding). The host has no visibility into the contents of the cbdata object, nor is permitted to alter it in any way.
+
+\ac{PMIx} callback functions typically return \code{void} - i.e., they do not return a value and therefore cannot indicate the result of an operation. Errors encountered by \ac{PMIx} callback functions are to be communicated via the \refapi{PMIx_Notify_event} \ac{API} using a status code appropriate to the error. Implementers and users wishing to receive such notifications must register a corresponding event handler via the \refapi{PMIx_Register_event_handler} \ac{API}.
+
```

Comment:
Good clarifications on the first 2 paragraphs.   I don't quite get the value of the 3rd paragraph.  The server module functions all return a pmix_status_t which they can use to communicate failures.  They sometimes have a callback function, such as pmix_modex_cbfunc_t, which returns a void, but these functions are only used to communicate the results of the server operation.  If the callback function back into the PMIx server fails, this is presumably a serious defect/failure within the server as there is no situation where we would ever expect that to happen.   If we are taking about the client side, then again I don't understand how the client provided callbacks would ever need to raise an event or why that would even be the right course of action.   Maybe I need a real example of what you are thinking here?

Reply:
I was responding to comments/questions from @rountree that were raised in the issue thread regarding "What should the error procedure be if cbdata is unrecognized?", which generalized into "what should the error procedure be if the callback function detects a problem?". Remember, the host does receive a number of callbacks from non-blocking function calls, and all of those callbacks return `void`. It could be a callback that returns the result of an operation, query of information or capabilities, etc.

The question shouldn't really be viewed as "how do I tell the PMIx library that something it gave me was wrong or indicated an error" - e.g., that a job control request failed. The question really is "how do I communicate that an error has occurred". Maybe I was looking for a specific capability and didn't find it, or I was trying to process the returned info and hit a data type that I don't recognize. Hence the use of `PMIx_Notify_event` - it lets both the target procs and the PMIx library know that something is wrong.

I can try to clarify it further based on the above explanation, if you find it helpful.

Reply:
I couldn't come up with a sensical way to explain it as your points were good, so I just removed the paragraph.